packages/sdk/src/oracle/store/has.ts

```diff
@@ -0,0 +1,21 @@
+import Read from "./read";
+import { ignoreError } from "../utils";
+import type { State } from "../types/state";
+
+// This class checks for existence for values you could potentially read. It mirrors the read interface but returns bools from functions.
+export default class Has {
+  private read: Read;
+  constructor(private state: State) {
+    // by design, reads do not mutate or cause side effects
+    this.read = new Read(state);
+  }
+  inputRequest = (): boolean => {
+    // the rule about reads is they will only throw errors if it cant return the value you request.
+    // If no errors are thrown the read was successful.
+    // Reads will never not return a value without throwing an error, enforced by TS.
+    return !!ignoreError(this.read.inputRequest);
```

Comment:
I think this just moves the problem.

The issue I see:
- There's an underlying bug in the code that's throwing.
- You do the existence check, that bug throws, so you never do the actual read, even though the data was there.
- You have an issue in your code, but it's invisible because the branch is being triggered by an error that wasn't the error you expected.

I think this is roughly the same as just catching and ignoring on the read because the _result_ is the same if there's an unexpected error: you don't do the read because of an error you didn't expect and the code doesn't tell you this unexpected error is being thrown.

Reply:
as per our offline convo, ive added a special type meant for throwing on if a value does not exist. Now the read class only throws these errors, and the `has` class only converts these errors into `false`, while letting the rest of the errors through. wrote unit tests to make sure this works as intended.

Reply:
Awesome, this is great!